Adding proposal for advancedcronjob

rishi-anand opened the pull request:
<!-- 
Please make sure you have read and understood the contributing guidelines;
https://github.com/openkruise/kruise/blob/master/CONTRIBUTING.md -->

### Ⅰ. Describe what this PR does
Implement CRD and controller for `AdvancedCronJob`

### Ⅱ. Does this pull request fix one issue?
<!--If so, add "fixes #xxxx" below in the next line, for example, fixes #15. Otherwise, add "NONE" -->
#215

### Ⅲ. List the added test cases (unit test/integration test) if any, please explain if no tests are needed.
This is proposal. Will add test cases in implementation.

### Ⅳ. Describe how to verify it
This is proposal. Will add more information in implementation PR.

### Ⅴ. Special notes for reviews
Discussion: https://github.com/openkruise/kruise/issues/215#issuecomment-715506813
Old PR for Proposal: https://github.com/openkruise/kruise/pull/427

@Fei-Guo, @FillZpp Please review

rishi-anand commented:
@Fei-Guo, Please review proposal in this PR.

Fei-Guo commented on docs/proposals/20201024-advancedcronjob.md:
````diff
@@ -0,0 +1,95 @@
+### User Stories
+
+- Implement a CRD which contains below fields and a controller which honors all the fields and reconciles accordingly.
+```
````
For API definition, can you add struct definition instead of CR yaml file?

I would suggest to introduce a JobTemplate struct that contains all the supported job types:
AdvancedCronJobSpec struct{
     template JobTemplate
}
type JobTemplate struct {
     JobTemplate *batchv1beta1.JobTemplateSpec
     BroadcastJobTemplate *BroadcastJobTemplateSpec
}

rishi-anand commented:
@Fei-Guo, I have added below template in spec and added api object in proposal.

```
// Specifies the job that will be created when executing a CronJob.
Template AdvancedCronJobTemplate `json:"template" protobuf:"bytes,7,opt,name=jobTemplate"`

type AdvancedCronJobTemplate struct {
	// Specifies the job that will be created when executing a CronJob.
	// +optional
	JobTemplate *batchv1beta1.JobTemplateSpec

	// Specifies the broadcastjob that will be created when executing a BroadcastCronJob.
	// +optional
	BroadcastJobTemplate *BroadcastJobTemplateSpec
}
```

Fei-Guo commented on docs/proposals/20201024-advancedcronjob.md:
```diff
@@ -0,0 +1,178 @@
+	Template AdvancedCronJobTemplate `json:"template" protobuf:"bytes,7,opt,name=jobTemplate"`
+}
+
+type AdvancedCronJobTemplate struct {
```
How about AdvancedCronJobTemplate -> CronJobTemplate?

rishi-anand replied:
Done in commit https://github.com/openkruise/kruise/pull/437/commits/5b0344eca29b2fc30c6a3f70846dc911779f6e3e
